xds/src/main/java/io/grpc/xds/XdsClientImpl.java

```diff
@@ -0,0 +1,721 @@
+  // always requesting for a singleton Listener resource, we only need to remember the
+  // a single Listener information.
+  // Upon receiving an LDS response, the entry for the RouteConfiguration indicated by the
+  // previous LDS response is deleted from RDS cache.
+  // Upon receiving an RDS response, each RouteConfiguration is inserted into RDS cache.
+
+  // Cached data for LDS responses, it is the RouteConfiguration name indicated by the requested
+  // Listener.
+  private String routeConfigNameInListener;
+  // Cached data for RDS responses, keyed by RouteConfiguration names.
+  // LDS responses indicate absent of RouteConfigurations and RDS responses indicate present
+  // of RouteConfigurations.
+  // Optimization: only cache clusterName filed in the RouteConfiguration messages of RDS
+  // responses.
+  private final Map<String, String> routeConfigNamesToClusterNames = new HashMap<>();
+
+  @Nullable
+  private AdsStream adsStream;
+  @Nullable
+  private BackoffPolicy retryBackoffPolicy;
+  @Nullable
+  private ScheduledHandle rpcRetryTimer;
+
+  // Following fields are set only after the ConfigWatcher registered. Once set, they should
+  // never change.
+  @Nullable
+  private ConfigWatcher configWatcher;
+  // The host name portion of "xds:" URI that the gRPC client targets for.
+  @Nullable
+  private String hostName;
+  // The "xds:" URI (including port suffix if present) that the gRPC client targets for.
+  @Nullable
+  private String ldsResourceName;
+
+  XdsClientImpl(
+      // URI of the management server to be connected to.
+      String serverUri,
+      Node node,
+      // List of channel credential configurations for the channel to management server.
+      // Should pick the first supported one.
+      List<ChannelCreds> channelCredsList,
+      SynchronizationContext syncContext,
+      ScheduledExecutorService timeService,
+      BackoffPolicy.Provider backoffPolicyProvider,
+      Stopwatch stopwatch) {
+    this(
+        buildChannel(checkNotNull(serverUri, "serverUri"),
+            checkNotNull(channelCredsList, "channelCredsList")),
+        node,
+        syncContext,
+        timeService,
+        backoffPolicyProvider,
+        stopwatch);
+  }
+
+  @VisibleForTesting
+  XdsClientImpl(
+      ManagedChannel channel,
+      Node node,
+      SynchronizationContext syncContext,
+      ScheduledExecutorService timeService,
+      BackoffPolicy.Provider backoffPolicyProvider,
+      Stopwatch stopwatch) {
+    this.channel = checkNotNull(channel, "channel");
+    this.node = checkNotNull(node, "node");
+    this.syncContext = checkNotNull(syncContext, "syncContext");
+    this.timeService = checkNotNull(timeService, "timeService");
+    this.backoffPolicyProvider = checkNotNull(backoffPolicyProvider, "backoffPolicyProvider");
+    this.stopwatch = checkNotNull(stopwatch, "stopwatch");
+  }
+
+  @Override
+  void shutdown() {
```

Comment:
protect against concurrent access? `synchronized` and have a flag to prevent repeat execution (e.g. `shutdownDone`)?

Reply:
Everything in this class is expected to run in a SynchronizationContext.

Reply:
Hmmm. So the `SynchronizationContext` is also passed to you (`this.syncContext`), is that the one? So:

- no comments/doc explain this particular aspect/restriction.
- instead of putting the onus on the caller of `shutdown` can you just use `this.syncContext` to execute the code inside?

I think it is good to protect your code against accidental misuse if easy to do so.